DownFall/BossController.swift

```diff
@@ -73,77 +106,177 @@ class BossController {
         case .newTurn:
             tiles = input.endTilesStruct
             advanceState()
+        case .transformation(let trans):
+            if let initialTrans = trans.first, case InputType.touch? = initialTrans.inputType,
+                let removed = initialTrans.tileTransformation?.first?.map({ $0.initial }) {
+                var newTargets: [TileCoord] = []
+                for coord in _currentTargets {
+                    if !removed.contains(coord) {
+                        newTargets.append(coord)
+                    }
+                }
+                _currentTargets = newTargets
+            }
         default: ()
         }
     }
     
-    enum BossAttack {
-        case column
-        case row
+    enum BossAttack: Hashable {
+        case hair(Int, Bool)
+        case destroy(Int, Bool)
         case spawn
         case bomb
     }
     
     func attack(basedOnRocks rocksEaten: [TileType]) -> [BossAttack] {
         
+        var attackedColumns = Set<Int>()
+        var attackedRows = Set<Int>()
+        
+        var attacksRow: Bool {
+            return Int.random(2) == 0
+        }
+        
+        func rowColumnAttack(_ attack: BossAttack) -> BossAttack {
+            let attacksRowOrColumn = attacksRow
+            let attackIndex: Int
+            if attacksRowOrColumn {
+                attackIndex = Int.random(boardSize, notInSet: attackedRows)
+                attackedRows.insert(attackIndex)
+            } else {
+                attackIndex = Int.random(boardSize, notInSet: attackedColumns)
+                attackedColumns.insert(attackIndex)
+            }
+            
+            switch attack {
+            case .hair:
+                return .hair(attackIndex, attacksRowOrColumn)
+            case .destroy:
+                return .destroy(attackIndex, attacksRowOrColumn)
+            default:
+                return attack
+            }
```

Comment:
Then in here we can match sure we we dont target the same columns/rows

Reply:
We could also figure out the TileCoords here as well for the row and column attacks.  We could then have then just work with an array of BossAttacks rather than a dictionary